### ISSUE TITLE
Implement `Debug` for `SessionStateBuilder

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#12555]
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/12550

While working on https://github.com/datafusion-contrib/datafusion-dft I found it would be nice to have Debug for [SessionStateBuilder](https://docs.rs/datafusion/latest/datafusion/execution/session_state/struct.SessionStateBuilder.html) so I could debug the actual state

### Describe the solution you'd like

I would like [SessionStateBuilder](https://docs.rs/datafusion/latest/datafusion/execution/session_state/struct.SessionStateBuilder.html) to implement Debug

### Describe alternatives you've considered

I started working on this and it turns out many of the sub structs (like `Analyzer` and `Optimizer` don't implement debug). This I recommend making a series of PRs to slowly add debug to the necessary structures so we can enable it for evertthing

**Comment by alamb** [comment]
A good way to find out what is left would be to put `#[derive(Debug)]` on SessionStateBuilder and then work down the errors

**Comment by alamb** [comment]
I think this is also a good first issue as it has a well defined scope and it is largely mechanical (doesn't require deep codebase knowledge)

Note I made a first step here: https://github.com/apache/datafusion/pull/12556

**Comment by alamb** [comment]
Here is another PR: https://github.com/apache/datafusion/pull/12557

**Comment by OussamaSaoudi** [comment]
take

**Comment by AnthonyZhOon** [comment]
Added a PR: #12624

**Comment by AnthonyZhOon** [comment]
I started on `CatalogProvider` but ran into `TableProvider` requiring Debug, waiting on #12557 which covers this to merge.

**Comment by AnthonyZhOon** [comment]
Finished being able to derive Debug on SessionStateBuilder #12632 but I believe it still needs to implement formatting to be more useful

**Comment by alamb** [comment]
> Finished being able to derive Debug on SessionStateBuilder https://github.com/apache/datafusion/pull/12632 but I believe it still needs to implement formatting to be more useful

Do you mean "derive a more usable `Display`" or something?

**Comment by AnthonyZhOon** [comment]
I think a pretty-print debug works okay using a `println!("{:#?}", state)`, most of the noise in the default was from printing long Vecs of scalar functions which have been moved to the end of debug now.
Noticed that the missing fields in `SessionState`'s `Debug` was from before we enabled it, so here's a pull request adding the remaining debug fields #12663 .
It's also possible to just derive Debug for both `SessionState` and `SessionStateBuilder` now.

**Comment by findepi** [comment]
Done in https://github.com/apache/datafusion/pull/12632. The issue can be closed.
thank you @AnthonyZhOon !
